socket/socket.js

```diff
@@ -0,0 +1,284 @@
+const User = require("../models/User").User;
+const Message = require("../models/Message");
+
+const socketEvents = (io) => {
+  io.on("connection", (socket) => {
+    /* Get the user's Chat list	*/
+    // go to messages collection, match userId to fromUserId or toUserId to to
+    // find who the current user has interacted with, extract all those userIds
+    // and send them back, indicating all contacts
+    socket.on(`chat-list`, async (data) => {
+      console.log(data);
+      console.log(socket.id);
+      if (data.userId == "") {
+        io.emit(`chat-list-response`, {
+          error: true,
+          message: "User not found",
+        });
+      } else {
+        try {
+          const [UserInfoResponse, chatlistResponse] = await Promise.all([
+            getUserInfo(data.userId),
+            getChatList(data.userId),
+          ]);
+
+          // this emit is returning the currentUser's chat list to themselves
+          io.to(socket.id).emit(`chat-list-response`, {
+            error: false,
+            singleUser: false,
+            chatList: chatlistResponse,
+          });
+
+          // this emit updates the status of any new online or offline users to everyone
+          // (useful for indicating which of your contacts is online/offline)
+          socket.broadcast.emit(`chat-list-response`, {
```

Comment:
I'm wondering if we should make this a separate broadcast for `user-login` or something. Right now I establish the socket connection when the user logs in (not when they open the chat) in the possible scenario of introducing more socket events like notifications and what not. Not 100% sure if that is the right idea, you can correct me if I'm wrong

Reply:
I thought of a similar thing but it'll involve changes to front-end code to actually have to emit extra events to update the status of user whereas before it sort of just happened alongside similar events (e.g. a logout just automatically tells everyone the user is gone, or "chat-list" is equivalent of log-on). I have made some changes and pushed to a separate branch called chat-2 in case you'd want to try. But yeah it'll involve more code on your side to have to emit the status-update manually and receive the success response and stuff

Reply:
for reference, it'd be something like this:
![image](https://user-images.githubusercontent.com/31676716/88830228-8412ce00-d182-11ea-9526-29c903fd13f4.png)